### ISSUE TITLE
Command substitution waits for backgrounded processes even if they are not attached to stdout

### ISSUE THREAD

**Issue by lexi-lambda** [fish-shell/fish-shell#5784]
In bash, command substitution only waits on backgrounded processes if they can possibly affect the result. This means that if they aren’t attached to stdout, substitution doesn’t wait. You can observe this by running the following command:

```bash
$ bash --version
GNU bash, version 4.4.23(1)-release (x86_64-apple-darwin17.5.0)
$ echo $({ sleep 3; echo delayed; } >&2 & echo immediate)
immediate
$ delayed
```

In fish, this doesn’t happen, and the substitution is forced to wait:

```fish
$ fish --version
fish, version 3.0.2
$ echo (begin sleep 3; echo delayed; end >&2 & echo immediate)
delayed
immediate
$
```

This is especially problematic in the context of process substitution. In bash, process substitution allows streaming results of the substituted process. You can observe that for yourself by running this command:

```bash
$ grep -F keyword <(echo uninteresting message; sleep 1; echo message with the keyword; sleep 1; echo another uninteresting message)
message with the keyword
```

The “message with the keyword” output will be printed after three seconds, the command will wait for another three, and then it will exit. But in fish, no output appears for the whole six seconds, then it appears all at once:

```fish
$ grep -F keyword (begin echo uninteresting message; sleep 3; echo message with the keyword; sleep 3; echo another uninteresting message; end | psub)
message with the keyword
```

If it were possible for background processes in subshells to be ignored, this could work fine, especially since `psub -F` does its writing in the background already.

**Comment by zanchey** [comment]
Blocks can't be backgrounded - see #238 (!) - so the first part of this issue is a duplicate of #238.

The `psub` behaviour, at least, is described in #1040 (and #4751) - it might be fixed although I'm not 100% clear.

I'll close this for now but speak up if I've misunderstood.
